common/djangoapps/util/help_desk.py

```diff
@@ -0,0 +1,443 @@
+    def __init__(self, msg, code, response):
+        self.msg = msg
+        self.error_code = code
+        self.response = response
+
+    def __str__(self):
+        return repr('%s: %s %s' % (self.error_code, self.msg, self.response))
+
+
+class _FreshdeskApi(object):
+    """
+    Freshdesk API implementation.
+    """
+
+    CACHE_PREFIX = 'FRESHDESK_API_CACHE'
+    CACHE_TIMEOUT = 60 * 60
+
+    def __init__(self):
+        """
+        Instantiate the Freshdesk API.
+
+        `HELPDESK_URL` and `HELPDESK_API_KEY` must be set
+        Set `HELPDESK` = 'Freshdesk'
+        in `django.conf.settings`.
+        """
+
+
+    def create_ticket(self, subject, desc, name, email, priority=1, status=2, group_id=None):
+        """
+        Create the given `ticket` in Freshdesk.
+
+        The ticket should have the format specified by the freshdesk api.
+        https://freshdesk.com/api#create_ticket
+        """
+        headers = {'Content-Type': 'application/json'}
+        payload = {
+            'subject': subject,
+            'description': desc,
+            'name': name,
+            'email': email,
+            'priority': priority,
+            'status': status,
+            'group_id': group_id
+        }
+
+        response = requests.post(
+            settings.HELPDESK_URL + '/api/v2/tickets',
+            auth=(settings.HELPDESK_API_KEY, "unused_but_required"),
+            headers=headers,
+            data=json.dumps(payload),
+            allow_redirects=False)
+
+        if response.ok:
+            message = json.loads(response.content)
+            return message['id']
+        else:
+            raise FreshdeskError(response.content, response.status_code, response)
+
+
+    def update_ticket(self, ticket_id, note, tags):
+        """
+        Update the Freshdeck ticket with id `ticket_id` using the given `note`.
+
+        The update should have the format specified by the freshdesk api.
+        https://freshdesk.com/api#add_note_to_a_ticket
+        """
+        headers = {'Content-Type': 'application/json'}
+        payload = {
+            "body" : note,
+            "private" : True
+        }
+
+        response = requests.post(
+            settings.HELPDESK_URL + '/api/v2/tickets/' + str(ticket_id) + '/notes',
+            auth=(settings.HELPDESK_API_KEY, "unused_but_required"),
+            headers=headers,
+            data=json.dumps(payload),
+            allow_redirects=False)
+
+        if response.ok:
+            return json.loads(response.content)
+        else:
+            raise FreshdeskError(response.content, response.status_code, response)
+
+
+    def get_group(self, name):
+        """
+        Find the Freshdesk group named `name`. Groups are cached for
+        CACHE_TIMEOUT seconds.
+
+        If a matching group exists, it is returned as a dictionary
+        with the format specifed by the freshdesk api.
+
+        Otherwise, returns None.
+        """
+        cache = caches['default']
```

Comment:
instead of import caches you can import cache `from django.core.cache import cache`
https://docs.djangoproject.com/en/2.2/topics/cache/#accessing-the-cache

Reply:
Great.